Add support for EST Clients to advertise support for automated cert p…

JamesGibo opened the pull request:
…rovisioning, using the NMOS Node API's caps attribute

Close #2

JamesGibo committed:
Add support for EST Clients to advertise support for automated cert provisioning,
using the NMOS Node API's caps attribute

andrewbonney commented on docs/1.0. Certificate Provisioning.md:
```diff
@@ -206,6 +206,17 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'caps' attribute of the Node API. Control interfaces can identify all Nodes which support BCP-003-03 by a key 'cert' and Uniform Resource Name (URN) having the base 'urn:x-nmos:control:cert' followed by a '/' character and the supported version.
```
```suggestion
If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'caps' attribute of the Node API. Control interfaces can identify all Nodes which support BCP-003-03 by a key 'cert' and Uniform Resource Name (URN) having the base 'urn:x-nmos:control:cert' followed by a '/' character and the supported version of this BCP.
```

We probably need a mechanism to cope with a client which supports multiple versions of the BCP. For Device 'controls' the array allows us to have multiple entries with different versioning.

andrewbonney reviewed (commented):
This warrants discussion, but we may find it's 'safer' to define a tag for this purpose rather than using 'caps' given that the caps object is currently under discussion more generally. If sufficiently generic, the tag could be used in the same way to indicate support for other NMOS specs.

It should probably be specified whether this means the Node supports EST and it is currently active, or whether this may mean that EST is supported but its current status may be enabled or disabled.

JamesGibo commented:
Adding Andrew's comment here so not lost from review
"We probably need a mechanism to cope with a client which supports multiple versions of the BCP. For Device 'controls' the array allows us to have multiple entries with different versioning."

JamesGibo committed:
Update docs/1.0. Certificate Provisioning.md

Co-authored-by: Andrew Bonney <andrew.bonney@bbc.co.uk>

JamesGibo committed:
Change advertisement to tags type

andrewbonney commented on docs/1.0. Certificate Provisioning.md:
```diff
@@ -206,13 +206,18 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
-"caps": {
-  "cert": "urn:x-nmos:control:cert/v1.0",
+"tags": {
+  "cert": {
```
```suggestion
  "urn:x-nmos:tag:cert": {
```

Matching the precedent in https://github.com/AMWA-TV/nmos-parameter-registers/blob/master/tags/grouphint.md a little more closely. The version could potentially make its way into the tag too.

JamesGibo committed:
Match the tags attribute elements to the precedent set by the
grouping API

https://github.com/AMWA-TV/nmos-parameter-registers/blob/master/tags/grouphint.md

prince-chrismc reviewed (commented):
Since I missed the call, was there a motivation to using an object with an `enabled` key over just using the URN as a string?

andrewbonney commented:
> Since I missed the call, was there a motivation to using an object with an `enabled` key over just using the URN as a string?

Were you thinking something like the following where true/false means enabled/disabled?
```
"urn:x-nmos:control:cert/v1.0": true
```

I realise given that we're looking at a tag this may actually have to be more like the following to match the IS-04 schemas, which is a little displeasing:
```
"urn:x-nmos:control:cert/v1.0": ["enabled"]
```

garethsb commented on docs/1.0. Certificate Provisioning.md:
```diff
@@ -206,6 +206,19 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the NMOS IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:control:cert/', followed by the supported version of this BCP. If EST is currently active is signalled using the 'enabled' boo ...
```
Nitpick. "AMWA IS-04" or just "IS-04", not "NMOS IS-04".

garethsb-sony commented:
Agree with Andrew, that it needs to fit the `"tags"` schema, which is a bit awkward for this purpose:

```
...
"tags": {
  ...
  "urn:x-nmos:tag:certprov/v1.0": ["enabled"] // or "true"?
}
...
```

🤔

JamesGibo committed:
Make EST tags advert match the tags schema

https://amwa-tv.github.io/nmos-discovery-registration/tags/v1.3/APIs/schemas/resolved/resource_core.html

JamesGibo commented:
I have updated the tag to match the tags schema and submitted a PR to nmos-parameter-registers to have the URN added.

https://github.com/AMWA-TV/nmos-parameter-registers/pull/20

prince-chrismc commented on docs/1.0. Certificate Provisioning.md:
````diff
@@ -206,6 +206,17 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
+```json
+...
+"tags": {
+  "urn:x-nmos:tag:certprov/v1.0": ["enabled"]
````
:thought_balloon: Might this be reasonable? keep one tag that has a list of supported versions, similar to the api pages?

```suggestion
  "urn:x-nmos:tag:certprov": ["v1.0"]
```

garethsb replied:
I was wondering about this! The only downside I see is that didn't someone ask about distinguishing between support for the spec and that support being currently active/enabled? But I think it's entirely reasonable for the resource to indicate the current configuration, so personally I like this proposal better.

prince-chrismc replied:
> The only downside I see is that didn't someone ask about distinguishing between support for the spec and that support being currently active/enabled?

I would explicitly say listed versions are active, if the tag or any version is missing then it is unsupported/disabled. Another talking point for the next call?

garethsb replied:
Your proposal makes perfect sense to me.

andrewbonney replied:
Yes, that makes sense to me too. We just need to make sure the text is clear that this shouldn't be present if the EST support is currently disabled by configuration.

JamesGibo replied:
Sounds good, I have updated the docs to reflect this

JamesGibo committed:
Make tags a list of enabled versions

andrewbonney commented on docs/1.0. Certificate Provisioning.md:
```diff
@@ -206,6 +206,19 @@ An EST Client MUST provide a method to replace the manufacturer-issued TLS Clien
 
 An EST Client MAY ignore all timestamps in the certificate validity periods during the certificate provisioning process if it does not accurately know the current time, see [Bootstrapping Remote Secure Key Infrastructures - Section 2.6.1](https://tools.ietf.org/html/draft-ietf-anima-bootstrapping-keyinfra-41#section-2.6.1) for more information.
 
+If an EST Client implements the AMWA IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:tag:certprov' and a list of enabled BCP-003-03 version. If BCP-003-03 is disabled the list MUST be empty. [NMOS Tags](https:// ...
```
```suggestion
If an EST Client implements the AMWA IS-04 Node API, then the EST Client SHOULD advertise that it supports BCP-003-03 using the 'tags' attribute of the Node API's `/self` endpoint. Control interfaces can identify all Nodes which support BCP-003-03 by a URN key 'urn:x-nmos:tag:certprov' and a list of enabled BCP-003-03 versions. If BCP-003-03 is disabled the list MUST be empty. This tag is defined in the NMOS Parameter Registers under [NMOS Tags](https://github.com/AMWA-TV/nmos-parameter-registers/tree/master/tags).
```

JamesGibo committed:
Update docs/1.0. Certificate Provisioning.md

Co-authored-by: Andrew Bonney <andrew.bonney@bbc.co.uk>

peterbrightwell commented:
Approved on call
